scaleway/resource_iot_hub_beta.go

```diff
@@ -0,0 +1,249 @@
+	return &schema.Resource{
+		Create: resourceScalewayIotHubBetaCreate,
+		Read:   resourceScalewayIotHubBetaRead,
+		Update: resourceScalewayIotHubBetaUpdate,
+		Delete: resourceScalewayIotHubBetaDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		SchemaVersion: 0,
+		Schema: map[string]*schema.Schema{
+			"enabled": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Wether to enable the hub or not",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the hub",
+			},
+			"product_plan": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "The product plan of the hub",
+				ValidateFunc: validation.StringInSlice([]string{
+					iot.ProductPlanPlanShared.String(),
+					iot.ProductPlanPlanDedicated.String(),
+					iot.ProductPlanPlanHa.String(),
+				}, false),
+			},
+
+			// Computed elements
+			"region":          regionSchema(),
+			"organization_id": organizationIDSchema(),
+			"created_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The date and time of the creation of the IoT Hub",
+			},
+			"updated_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The date and time of the last update of the IoT Hub",
+			},
+			"status": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The status of the hub",
+			},
+			"endpoint": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The enpoint to connect the devices to",
+			},
+			"device_count": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The number of registered devices in the Hub",
+			},
+			"connected_device_count": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The current number of connected devices in the Hub",
+			},
+		},
+	}
+}
+
+func resourceScalewayIotHubBetaCreate(d *schema.ResourceData, m interface{}) error {
+	iotAPI, region, err := iotAPIWithRegion(d, m)
+	if err != nil {
+		return err
+	}
+
+	////
+	// Create cluster
+	////
+
+	req := &iot.CreateHubRequest{
+		Region:         region,
+		Name:           expandOrGenerateString(d.Get("name"), "hub"),
+		OrganizationID: d.Get("organization_id").(string),
+		ProductPlan:    iot.ProductPlan(d.Get("product_plan").(string)),
+	}
+
+	res, err := iotAPI.CreateHub(req)
+	if err != nil {
+		return err
+	}
+
+	err = waitIotHub(iotAPI, region, res.ID, iot.HubStatusReady)
+	if err != nil {
+		return err
+	}
+
+	// Enable/Disable hub if needed. The only case we need to check is to eventually
+	// disable the hub (enabled by default)
+	// We need to ensure the hub was fully enabled before because the hub cannot
+	// be updated while enabling.
+	if enabled, ok := d.GetOk("enabled"); ok && !enabled.(bool) {
```

Comment:
To avoid code duplication you could consider. Not sure about the `HasChange` method comportement thought
```
d.SetId(newRegionalId(region, res.ID))
resourceScalewayIotHubBetaRead(d, m)
return resourceScalewayIotHubBetaUpdate(d, m)
```

Reply:
Calling the update method might probably trigger an enable on an already enabled hub